src/solver.jl

```diff
@@ -1,159 +1,67 @@
-- If `show_count` is true, returns a tuple `(mis, count)` where `mis` is the maximum independent set (MIS) of the graph `g` and `count` is the number of iterations performed during the solving process.
-"""
-function missolve(g::SimpleGraph; show_count = false, strategy::BranchingStrategy = NaiveBranching(), kneighbor::Int = 2, use_rv::Bool = true, use_filter::Bool = true)
-    mis = mis_solver(g, strategy, kneighbor, use_rv, use_filter)
+function missolve(g::SimpleGraph; show_count = false, strategy::AbsractBranching = NaiveBranching(), measurement::AbstractMeasurement = NaiveMeasure(), vertex_select::AbstractVertexSelector = MinBoundSelector(2), filter::AbstractTruthFilter = EnvFilter())
+    mis = mis_solver(g, strategy, measurement, vertex_select, filter)
     return show_count ? (mis.mis, mis.count) : mis.mis
 end
 
-function rv(vertices::Vector{Int}, g::SimpleGraph, clause::Clause{N}) where N
-    removed_vertices = Int[]
-    for (k, v) in enumerate(vertices)
-        if readbit(clause.mask, k) == 1
-            push!(removed_vertices, v)
-            if readbit(clause.val, k) == 1
-                append!(removed_vertices, neighbors(g, v))
-            end
-        end
-    end
-    return unique!(removed_vertices)
-end
-
-function mis_solver(g::SimpleGraph, strategy::BranchingStrategy, kneighbor::Int, use_rv::Bool, use_filter::Bool)
+function mis_solver(g::SimpleGraph, strategy::AbsractBranching, measurement::AbstractMeasurement, vertex_select::AbstractVertexSelector, filter::AbstractTruthFilter)
     dg = degree(g)
     if nv(g) == 0 || nv(g) == 1
         return CountingMIS(nv(g))
     elseif (0 ∈ dg) || (1 ∈ dg)
         v = findfirst(x -> (x==0)||(x==1), dg)
-        return 1 + mis_solver(induced_subgraph(g, setdiff(1:nv(g), v ∪ neighbors(g, v)))[1], strategy, kneighbor, use_rv, use_filter)
+        return 1 + mis_solver(induced_subgraph(g, setdiff(1:nv(g), v ∪ neighbors(g, v)))[1], strategy,measurement, vertex_select, filter)
+    elseif (2 ∈ dg)
+        v = findfirst(x -> (x==2), dg)
+        return folding(g, v, strategy, measurement, vertex_select, filter)
     elseif maximum(dg) ≥ 6
         v = argmax(dg)
-        return max(1 + mis_solver(induced_subgraph(g, setdiff(1:nv(g), v ∪ neighbors(g, v)))[1], strategy, kneighbor, use_rv, use_filter), mis_solver(induced_subgraph(g, setdiff(1:nv(g), v))[1], strategy, kneighbor, use_rv, use_filter))
+        return max(1 + mis_solver(induced_subgraph(g, setdiff(1:nv(g), v ∪ neighbors(g, v)))[1], strategy, measurement, vertex_select, filter), mis_solver(induced_subgraph(g, setdiff(1:nv(g), v))[1], strategy, measurement, vertex_select, filter))
     else
-        vertices, openvertices, dnf = optimal_branching_dnf(g, strategy, kneighbor, use_rv, use_filter)
-        @assert !isempty(vertices)
+        vertices, openvertices, dnf, γ = optimal_branching_dnf(g, strategy, measurement, vertex_select, filter)
```

Comment:
Please return a struct, like `Branches`, instead of a quadruple. This struct should contain a vector of branches, and each branch contains removed vertices and size reduction (under certain measure).

Implement a function `effective_γ` on it.

Reply:
Two new structures are defined to store the branching rules:
```julia
struct Branch{T}
    vertices_removed::Vector{Int}
    size_reduced::T
    mis::Int
end

struct Branches{T}
    branches::Vector{Branch{T}}
end

function effective_γ(branches::Branches{T}) where{T}
    return complexity([branches.branches[i].size_reduced for i in 1:length(branches.branches)])
end
```
The `Branch` now store the vertices removed, size reduced under certain measurement and the mis size of the subgraph.
In that case maybe we can just remove the `DNF`?

Reply:
It is ok to have both. `DNF` is a good abstraction.